### PR TITLE
JBIDE-21447 - Create OpenShift Server Adapter from service/route context menu

### PR DESCRIPTION
**Pull request by xcoulon**
Refactored the ServerSettingsWizardFragment to extract the widgets
 into ServerSettingsWizardPage and keep the WizardFragment that now
 includes the WizardPage via a wrapper.
Added context menu entries in the OpenShift Explorer view and in the
 Properties view
Exported the "org.jboss.tools.openshift.core.util" package to be able to use
 it in the org.jboss.tools.openshift.ui project.

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: plugins/org.jboss.tools.openshift.core/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -32,6 +32,7 @@ Export-Package: org.jboss.tools.openshift.core;x-friends:="org.jboss.tools.opens
  org.jboss.tools.openshift.core.connection;x-friends:="org.jboss.tools.openshift.test,org.jboss.tools.openshift.ui",
  org.jboss.tools.openshift.core.preferences;x-friends:="org.jboss.tools.openshift.test,org.jboss.tools.openshift.ui",
  org.jboss.tools.openshift.core.server;x-friends:="org.jboss.tools.openshift.test,org.jboss.tools.openshift.ui",
+ org.jboss.tools.openshift.core.util,
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/META-INF/MANIFEST.MF]: ;x-friends:="org.jboss.tools.openshift.test,org.jboss.tools.openshift.ui",

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java [hunk 2]
```diff
@@ -166,531 +97,70 @@ public boolean hasComposite() {
...
-
-		// browse projects
-		Button browseProjectsButton = new Button(container, SWT.NONE);
-		browseProjectsButton.setText("Browse...");
-		GridDataFactory.fillDefaults()
-				.align(SWT.LEFT, SWT.CENTER)
-				.hint(100, SWT.DEFAULT)
-				.applyTo(browseProjectsButton);
-		browseProjectsButton.addSelectionListener(onBrowseProjects(model, browseProjectsButton.getShell()));
+	@Override
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java]: useless method

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java [hunk 3]
```diff
@@ -166,531 +97,70 @@ public boolean hasComposite() {
...
-					Object selectedProject = dialog.getFirstResult();
-					if (selectedProject instanceof IProject) {
-						model.setDeployProject((org.eclipse.core.resources.IProject) selectedProject);
-					}
-				}
-			}
-		};
+	@Override
+	public void setComplete(boolean complete) {
+		super.setComplete(complete);
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java]: useless method

**Comment by fbricon** [comment]
Seems to work (http://screencast.com/t/sGetXm8hl28), but I have a few remarks, ordered by priority 
1 - When opening the wizard, the service selected in the OpenShift Explorer should also be selected in the wizard.
2 - Empty default route section?
3 - When the server already exists but the servers view is closed, opening the Server Adapter menu opens the Servers view but doesn't select the matching server
4 - I fear the `Server Adapter` is a bit confusing, as it's an WTP internal concept. I can't find something more explicit/user friendly for now, but will try to think of something.
5 - Ideally the Workspace project, shared with git and having a remote repo url identical to the selected service's build config git url should be preselected in the wizard. But that's icing on the cake, we could have that in a later commit.

**Comment by xcoulon** [comment]
thanks for your review, @fbricon. I'll work on it and submit an update of the PR.

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java (hunk 3 above)
Comment (xcoulon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java]: right

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java (hunk 2 above)
Comment (xcoulon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java]: right

File: plugins/org.jboss.tools.openshift.core/META-INF/MANIFEST.MF (hunk 1 above)
Comment (xcoulon) [inline plugins/org.jboss.tools.openshift.core/META-INF/MANIFEST.MF]: thanks for spotting this. I used a Quick-Fix to include this package in the 'exports' but I did not realize that the 'x-friends' option was missing.

**Comment by xcoulon** [comment]
@fbricon @adietish please review this PR. All comments from Fred were taken into account (even the last one) except the "Server Adapter.." command name for now. 
I used a few lambda expressions to find the service in the serviceItems in the model, using a new utility class to flatten the ObservableTreeItem hierarchies. I think it makes the code simpler to read. Your feedback is welcome.

**Comment by xcoulon** [comment]
@fbricon @adietish pull request was rebased againt `jbosstools-4.3.x` branch to include the latest changes. It also fixes bugs when the user calls the command from a route or a service in the properties view.
Remained unchanged is the "Server Adapter..." label, but we could fix that in a subsequent commit if necessary

**Comment by fbricon** [comment]
- So I don't know why yet but it takes between 30sec to 1 min to open the server wizard after clicking on the menu. This is definitely not user friendly
- It seems the code properly detects the  workspace project matching the git repo defined in the build config (when looking at instance variables set in the ServerSettingsWizardPage constructor ), but this is not the project preselected in the UI.
I believe this is unrelated, but when running against console.engint.openshift.com, all my projects/services error with : 
```
!ENTRY org.eclipse.core.jobs 4 2 2016-02-26 12:37:05.170
!MESSAGE An internal error occurred during: "Loading projects and services...".
!STACK 0
com.openshift.restclient.OpenShiftException: Could not get resource nodejs-example:latest in namespace nonode: {
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "imageStreamTag \"nodejs-example:latest\" not found",
  "reason": "NotFound",
  "details": {
    "name": "nodejs-example:latest",
    "kind": "imageStreamTag"
  },
  "code": 404
}

	at com.openshift.internal.restclient.DefaultClient.createOpenShiftException(DefaultClient.java:482)
	at com.openshift.internal.restclient.DefaultClient.get(DefaultClient.java:306)
	at org.jboss.tools.openshift.core.connection.Connection.getResource(Connection.java:387)
	at org.jboss.tools.openshift.internal.ui.models.DeploymentResourceMapper.lambda$5(DeploymentResourceMapper.java:106)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1540)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.jboss.tools.openshift.internal.ui.models.DeploymentResourceMapper.getImageStreamTagsFor(DeploymentResourceMapper.java:107)
	at org.jboss.tools.openshift.internal.ui.models.DeploymentResourceMapper.lambda$2(DeploymentResourceMapper.java:87)
	at java.util.stream.Collectors.lambda$toMap$173(Collectors.java:1321)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1540)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.jboss.tools.openshift.internal.ui.models.DeploymentResourceMapper.getAllImageStreamTags(DeploymentResourceMapper.java:85)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel$ProjectImageStreamTags.createImageStreamsMap(ServerSettingsViewModel.java:334)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel$ProjectImageStreamTags.<init>(ServerSettingsViewModel.java:327)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.lambda$1(ServerSettingsViewModel.java:279)
	at java.util.stream.Collectors.lambda$toMap$173(Collectors.java:1321)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.createImageStreamTagsMap(ServerSettingsViewModel.java:277)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.loadResources(ServerSettingsViewModel.java:261)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.loadResources(ServerSettingsViewModel.java:255)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsWizardPage$2.run(ServerSettingsWizardPage.java:244)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: com.openshift.internal.restclient.http.NotFoundException: {
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "imageStreamTag \"nodejs-example:latest\" not found",
  "reason": "NotFound",
  "details": {
    "name": "nodejs-example:latest",
    "kind": "imageStreamTag"
  },
  "code": 404
}

	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.createException(UrlConnectionHttpClient.java:230)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:165)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:141)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.get(UrlConnectionHttpClient.java:103)
	at com.openshift.internal.restclient.DefaultClient.get(DefaultClient.java:302)
	... 36 more
Caused by: java.io.FileNotFoundException: https://console.engint.openshift.com/oapi/v1/namespaces/nonode/imagestreamtags/nodejs-example:latest
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1836)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:161)
	... 39 more

```

**Comment by adietish** [comment]
@fbricon The 404 that you get is what we try to fix in https://issues.jboss.org/browse/JBIDE-21731

**Comment by xcoulon** [comment]
@fbricon here I have to wait for a few seconds before the wizard appears, and I suspect this is due to the resources loading. But it certainly does not take 30s to 1 min... Maybe we could try to load the resources in parallel ? @adietish, do you think it's doable ?
In the mean time, the time to load the resources is temporary measured and printed in the errors view (with level `INFO`). @fbricon can you let me know what value you have (here it varies between 800ms and 1100ms).
Also, I updated the PR with a fix to the selected project, please let me know if it works.

**Comment by adietish** [comment]
@xcoulon yes we can at least load the services in parallel (currently the wizard is waiting for the loading job to finish - WizardUtils#runInWizard(job)). On a technical level the list of services is using databinding and is capable of lazy loading.

**Comment by fbricon** [comment]
See http://screencast.com/t/jweIzStNk

**Comment by xcoulon** [comment]
@fbricon okay.... may I ask if you had the same issue when opening the Server Creation wizard from the Servers view *without* my patch applied ? Also, do you have a lot of connections and OS projects to load ?

**Comment by adietish** [comment]
@fbricon @xcoulon I agree that loading takes a bit too much time to excel at user friendlyness. The loading shall be changed in https://issues.jboss.org/browse/JBIDE-21731 and I thus would like to merge this regardless of the current loading times.

**Comment by adietish** [comment]
@fbricon @xcoulon I agree that the initial loading takes too much time. Loading will most likely be changed to be more lazy (less eager) in https://issues.jboss.org/browse/JBIDE-21731. I thus want to merge this without any change to the loading strategy.

**Comment by adietish** [comment]
thanks!!
merged into jbosstools-4.3.x and cherry-picked into master

**Comment by xcoulon** [comment]
thanks ;-)

**Pull request merged**